### ISSUE TITLE
Remove `discovery.type` BWC layer from the EC2/Azure/GCE plugins

### ISSUE THREAD

**Issue by bleskes** [elastic/elasticsearch#24543]
Those plugins don't replace the discovery logic but rather only provide a custom unicast host provider for their respective platforms. in 5.1 we introduced the  `discovery.zen.hosts_provider` setting to better reflect it. We should remove the [BWC code](https://github.com/elastic/elasticsearch/blob/master/plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/Ec2DiscoveryPlugin.java#L105) in those plugins.

**Comment by Jefferysan** [comment]
Hi, I am new to the project and would like to take this as my first time contribution.

**Comment by bleskes** [comment]
@Jefferysan thank you for offering to contribute. Before starting I want to double check you have read our [contributing guide](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-code-and-documentation-changes). Since your github profile is empty I will go out on a limb and also ask you pay special attention to the section about contributing about a class assignment. Once you got everything working and have successfully run the tests, we can start iterating on the code. If you have any more questions, please feel free to ask.

PS this is marked as a blocker for 6.0. This means that there might be a point where we need it fixed. If you think you might not be able to drive this to a conclusion quickly (i.e., go through reviews and update your code accordingly until this is merged), maybe this one is not a good way to start.

**Comment by Jefferysan** [comment]
It had the low hanging fruit label attached to the issue so I thought it would be a good place to start. Is it maybe not the case?

**Comment by bleskes** [comment]
> It had the low hanging fruit label attached to the issue so I thought it would be a good place to start. Is it maybe not the case?

My comments were about setting up a working environment so you'll need that for any PR. I think this one is OK to start with. I just wanted make sure it might become time sensitive at some point, so if you're not sure you will be able to finish it soon, you should probably pickup something else.

**Comment by jmorph99** [comment]
Hi, I have forked the source.  This error looks simple, remove the values from those three files and any test files.  I installed this on Ubuntu, I should probably document that!  This took me a while.  First I started with NetBeans ...   So I have to use a new compiler now so I am trying IntelliJ.  Anyways, my testClasses tests run fine but my integTest tests error, Do I need that to be working in order to do this project?  If not I think I can follow your instructions to submit the fixes requested, I do not need a lot of hand-holding.

**Comment by rjernst** [comment]
> my integTest tests error, Do I need that to be working in order to do this project

Yes, running `gradle check` needs to pass before any PR can be merged.

**Comment by jmorph99** [comment]
Yes that errors - I will execute it again with the "--stacktrace" option
and output it to a text file.  If you could take a quick look at it I think
it might help.  I will send it later, the tests take a long time.

**Comment by jmorph99** [comment]
Also, I haven't done any modifications to the source code yet, I am simply
trying to execute the tests after the Gradle install.

**Comment by jmorph99** [comment]
Ok, I eventually just reran it and for no apparent reason, it completed.



*BUILD SUCCESSFULTotal time: 34 mins 52.896 secs*

Onward I go,
John

**Comment by jmorph99** [comment]
I think I just need to remove the overwritten getDiscoveryTypes function from these three files.  Am I correct?

@Override
    public Map<String, Supplier<Discovery>> getDiscoveryTypes(ThreadPool threadPool, TransportService transportService,
                                                              NamedWriteableRegistry namedWriteableRegistry,
                                                              MasterService masterService, ClusterApplier clusterApplier,
                                                              ClusterSettings clusterSettings, UnicastHostsProvider hostsProvider,
                                                              AllocationService allocationService) {
        // this is for backcompat with pre 5.1, where users would set discovery.type to use ec2 hosts provider
        return Collections.singletonMap(EC2, () ->
            new ZenDiscovery(settings, threadPool, transportService, namedWriteableRegistry, masterService, clusterApplier,
                clusterSettings, hostsProvider, allocationService));
    }

**Comment by rjernst** [comment]
@jmorph99 Yes that sounds about right.

**Comment by rjernst** [comment]
I believe some of the overridden `additionalSettings()` can also be removed, which injected the appropriate new setting when detecting the old setting.

**Comment by jmorph99** [comment]
I am still working on this.  I have to learn a new build tool (gradle), a
new IDE (intelliJ) and a new code base(Elasticsearch), but I will deliver.

**Comment by bleskes** [comment]
Thanks for picking this up. Let us know if you need any help.

**Comment by jmorph99** [comment]
Thanks Boaz,

After removing the override I get a test error.  I am not sure if the
modification is bad or I need to modify/remove the test.  Could you provide
me with guidance?



John

Testing started at 6:17 PM ...
6:17:17 PM: Executing external task 'check'...
:buildSrc:compileJava UP-TO-DATE
:buildSrc:compileGroovy UP-TO-DATE
:buildSrc:writeVersionProperties UP-TO-DATE
:buildSrc:processResources UP-TO-DATE
:buildSrc:classes UP-TO-DATE
:buildSrc:jar UP-TO-DATE
:buildSrc:assemble UP-TO-DATE
:buildSrc:compileTestJava UP-TO-DATE
:buildSrc:compileTestGroovy NO-SOURCE
:buildSrc:processTestResources NO-SOURCE
:buildSrc:testClasses UP-TO-DATE
:buildSrc:test NO-SOURCE
:buildSrc:check UP-TO-DATE
:buildSrc:build UP-TO-DATE
=======================================
Elasticsearch Build Hamster says Hello!
=======================================
  Gradle Version        : 3.5
  OS Info               : Linux 4.4.0-78-generic (amd64)
  JDK Version           : Oracle Corporation 1.8.0_111 [Java HotSpot(TM)
64-Bit Server VM 25.111-b14]
  JAVA_HOME             : /usr/lib/jvm/jdk1.8.0_111
The executable property on ForkOptions has been deprecated and is scheduled
to be removed in Gradle 5.0. Please use javaHome instead.
Could not find vagrant: A problem occurred starting process 'command
'vagrant''
secured hdfsFixture is unsupported, please install Vagrant 1.8.6+ to enable
:core:compileJava UP-TO-DATE
:core:generateModulesList UP-TO-DATE
:core:generatePluginsList UP-TO-DATE
:core:processResources UP-TO-DATE
:core:classes UP-TO-DATE
:core:jar UP-TO-DATE
:plugins:discovery-gce:compileJava
:plugins:discovery-gce:compileJava - is not incremental (e.g. outputs have
changed, no previous execution, etc.).
:plugins:discovery-gce:processResources NO-SOURCE
:plugins:discovery-gce:classes
:plugins:discovery-gce:copyCheckstyleConf UP-TO-DATE
:plugins:discovery-gce:checkstyleMain
:client:rest:compileJava UP-TO-DATE
:client:rest:processResources NO-SOURCE
:client:rest:classes UP-TO-DATE
:client:rest:jar UP-TO-DATE
:plugins:discovery-gce:copyPluginPropertiesTemplate
:plugins:discovery-gce:pluginProperties UP-TO-DATE
:test:framework:compileJava UP-TO-DATE
:test:framework:processResources UP-TO-DATE
:test:framework:classes UP-TO-DATE
:test:framework:jar UP-TO-DATE
:plugins:discovery-gce:compileTestJava
:plugins:discovery-gce:createKey
:plugins:discovery-gce:processTestResources
:plugins:discovery-gce:testClasses
:plugins:discovery-gce:checkstyleTest
:plugins:discovery-gce:forbiddenApisMain
:plugins:discovery-gce:forbiddenApisTest
:plugins:discovery-gce:forbiddenApis
:plugins:discovery-gce:generateNotice UP-TO-DATE
:plugins:discovery-gce:jar
:plugins:discovery-gce:bundlePlugin
:distribution:buildCoreNotice UP-TO-DATE
:modules:transport-netty4:compileJava UP-TO-DATE
:modules:transport-netty4:processResources NO-SOURCE
:modules:transport-netty4:classes UP-TO-DATE
:modules:transport-netty4:jar UP-TO-DATE
:modules:transport-netty4:copyPluginPropertiesTemplate
:modules:transport-netty4:pluginProperties UP-TO-DATE
:modules:transport-netty4:bundlePlugin UP-TO-DATE
:distribution:buildTransportModules UP-TO-DATE
:distribution:integ-test-zip:createPluginsDir
:distribution:tools:java-version-checker:compileJava UP-TO-DATE
:distribution:tools:java-version-checker:processResources NO-SOURCE
:distribution:tools:java-version-checker:classes UP-TO-DATE
:distribution:tools:java-version-checker:jar UP-TO-DATE
:distribution:tools:plugin-cli:compileJava UP-TO-DATE
:distribution:tools:plugin-cli:processResources NO-SOURCE
:distribution:tools:plugin-cli:classes UP-TO-DATE
:distribution:tools:plugin-cli:jar UP-TO-DATE
:distribution:integ-test-zip:buildZip UP-TO-DATE
:plugins:discovery-gce:integTestCluster#prepareCluster.cleanShared
:plugins:discovery-gce:integTestCluster#clean
:plugins:discovery-gce:integTestCluster#checkPrevious SKIPPED
:plugins:discovery-gce:integTestCluster#stopPrevious SKIPPED
:plugins:discovery-gce:integTestCluster#extract
:plugins:discovery-gce:integTestCluster#configure
:plugins:discovery-gce:integTestCluster#copyPlugins
:plugins:discovery-gce:integTestCluster#installDiscoveryGcePlugin
:plugins:discovery-gce:integTestCluster#start
:rest-api-spec:compileJava NO-SOURCE
:rest-api-spec:processResources UP-TO-DATE
:rest-api-spec:classes UP-TO-DATE
:rest-api-spec:jar UP-TO-DATE
:plugins:discovery-gce:copyRestSpec UP-TO-DATE
:plugins:discovery-gce:checkstyle
:plugins:discovery-gce:dependencyLicenses
:plugins:discovery-gce:forbiddenPatterns UP-TO-DATE
:plugins:discovery-gce:jarHell
:plugins:discovery-gce:licenseHeaders
:test:logger-usage:compileJava UP-TO-DATE
:test:logger-usage:processResources NO-SOURCE
:test:logger-usage:classes UP-TO-DATE
:test:logger-usage:jar UP-TO-DATE
:plugins:discovery-gce:loggerUsageCheck
:build-tools:compileJava UP-TO-DATE
:build-tools:compileGroovy UP-TO-DATE
:build-tools:writeVersionProperties UP-TO-DATE
:build-tools:processResources UP-TO-DATE
:build-tools:classes UP-TO-DATE
:build-tools:jar UP-TO-DATE
:plugins:discovery-gce:namingConventions
:plugins:discovery-gce:thirdPartyAudit UP-TO-DATE
:plugins:discovery-gce:precommit
:plugins:discovery-gce:test
==> Test Info: seed=3D8ACCBEE4BD7E73; jvms=3; suites=4
Suite: org.elasticsearch.discovery.gce.GceDiscoverTests
  1> [2017-05-30T22:17:34,644][INFO ][o.e.d.g.GceDiscoverTests ]
[testJoin]: before test
  1> [2017-05-30T22:17:34,647][INFO ][o.e.d.g.GceDiscoverTests ]
[GceDiscoverTests#testJoin]: setting up test
  1> [2017-05-30T22:17:34,680][INFO ][o.e.t.InternalTestCluster] Setup
InternalTestCluster
[SUITE-CHILD_VM=[0]-CLUSTER_SEED=[2818117216213349196]-HASH=[2887AC9BF9AD]-cluster]
with seed [271BF6AD57CF334C] using [0] dedicated masters, [2] (data) nodes
and [0] coord only nodes (min_master_nodes are [auto-managed])
  1> [2017-05-30T22:17:34,869][INFO ][o.e.n.Node               ] [node_s0]
initializing ...
  1> [2017-05-30T22:17:34,901][INFO ][o.e.e.NodeEnvironment    ] [node_s0]
using [1] data paths, mounts [[/ (/dev/sda1)]], net usable_space [80gb],
net total_space [204.2gb], types [ext4]
  1> [2017-05-30T22:17:34,901][INFO ][o.e.e.NodeEnvironment    ] [node_s0]
heap size [491mb], compressed ordinary object pointers [true]
  1> [2017-05-30T22:17:34,902][INFO ][o.e.n.Node               ] [node_s0]
node name [node_s0], node ID [qXhJ4YgcR4us48OJOcU2-g]
  1> [2017-05-30T22:17:34,902][INFO ][o.e.n.Node               ] [node_s0]
version[6.0.0-alpha2-SNAPSHOT], pid[20900],
build[d1318e4/2017-05-30T10:04:07.753Z], OS[Linux/4.4.0-78-generic/amd64],
JVM[Oracle Corporation/Java HotSpot(TM) 64-Bit Server
VM/1.8.0_111/25.111-b14]
  1> [2017-05-30T22:17:34,902][INFO ][o.e.n.Node               ] [node_s0]
JVM arguments [-ea, -esa, -Xmx512m, -Xms512m,
-XX:+HeapDumpOnOutOfMemoryError,
-XX:HeapDumpPath=/home/murphy/elastic/elasticsearch/plugins/discovery-gce/build/heapdump,
-Dtests.prefix=tests, -Dtests.seed=3D8ACCBEE4BD7E73,
-Dtests.security.manager=false, -Djava.io.tmpdir=./temp,
-Dtests.gradle=true, -Dtests.artifact=discovery-gce,
-Djava.awt.headless=true, -Djdk.io.permissionsUseCanonicalPath=true,
-Dtests.ifNoTests=fail, -Dtests.task=:plugins:discovery-gce:test,
-Djna.nosys=true, -Dtests.logger.level=WARN,
-Djunit4.childvm.cwd=/home/murphy/elastic/elasticsearch/plugins/discovery-gce/build/testrun/test/J0,
-Djunit4.childvm.id=0, -Djunit4.childvm.count=3]
  1> [2017-05-30T22:17:34,902][WARN ][o.e.n.Node               ] [node_s0]
version [6.0.0-alpha2-SNAPSHOT] is a pre-release version of Elasticsearch
and is not suitable for production
  1> [2017-05-30T22:17:34,912][INFO ][o.e.p.PluginsService     ] [node_s0]
no modules loaded
  1> [2017-05-30T22:17:34,913][INFO ][o.e.p.PluginsService     ] [node_s0]
loaded plugin [org.elasticsearch.discovery.gce.GceDiscoverTests$TestPlugin]
  1> [2017-05-30T22:17:34,913][INFO ][o.e.p.PluginsService     ] [node_s0]
loaded plugin [org.elasticsearch.plugin.discovery.gce.GceDiscoveryPlugin]
  1> [2017-05-30T22:17:34,913][INFO ][o.e.p.PluginsService     ] [node_s0]
loaded plugin [org.elasticsearch.test.ESIntegTestCase$TestSeedPlugin]
  1> [2017-05-30T22:17:34,913][INFO ][o.e.p.PluginsService     ] [node_s0]
loaded plugin [org.elasticsearch.test.discovery.TestZenDiscovery$TestPlugin]
  1> [2017-05-30T22:17:34,913][INFO ][o.e.p.PluginsService     ] [node_s0]
loaded plugin [org.elasticsearch.transport.MockTcpTransportPlugin]
  1> [2017-05-30T22:17:34,913][WARN ][o.e.d.p.d.g.GceDiscoveryPlugin] Using
discovery.type setting to set hosts provider is deprecated. Set
"discovery.zen.hosts_provider: gce" instead
  1> [2017-05-30T22:17:35,647][INFO ][o.e.d.g.GceDiscoverTests ]
[GceDiscoverTests#testJoin]: cleaning up after test
  1> [2017-05-30T22:17:35,656][INFO ][o.e.d.g.GceDiscoverTests ]
[testJoin]: after test
  2> REPRODUCE WITH: gradle :plugins:discovery-gce:test
-Dtests.seed=3D8ACCBEE4BD7E73
-Dtests.class=org.elasticsearch.discovery.gce.GceDiscoverTests
-Dtests.method="testJoin" -Dtests.security.manager=false
-Dtests.locale=ar-MA -Dtests.timezone=Africa/Nouakchott
  2> REPRODUCE WITH: gradle :plugins:discovery-gce:test
-Dtests.seed=3D8ACCBEE4BD7E73
-Dtests.class=org.elasticsearch.discovery.gce.GceDiscoverTests
-Dtests.method="testJoin" -Dtests.security.manager=false
-Dtests.locale=ar-MA -Dtests.timezone=Africa/Nouakchott
ERROR   1.08s J0 | GceDiscoverTests.testJoin <<< FAILURES!
   > Throwable #1: java.lang.IllegalArgumentException: Unknown discovery
type [gce]
   >     at
org.elasticsearch.discovery.DiscoveryModule.<init>(DiscoveryModule.java:102)
   >     at org.elasticsearch.node.Node.<init>(Node.java:438)
   >     at org.elasticsearch.node.MockNode.<init>(MockNode.java:60)
   >     at
org.elasticsearch.test.InternalTestCluster.buildNode(InternalTestCluster.java:608)
   >     at
org.elasticsearch.test.InternalTestCluster.reset(InternalTestCluster.java:1040)
   >     at
org.elasticsearch.test.InternalTestCluster.beforeTest(InternalTestCluster.java:978)
   >     at
org.elasticsearch.test.ESIntegTestCase.beforeInternal(ESIntegTestCase.java:373)
   >     at
org.elasticsearch.test.ESIntegTestCase.setupTestCluster(ESIntegTestCase.java:2044)
   >     at java.lang.Thread.run(Thread.java:745)Throwable #2:
java.lang.RuntimeException: java.nio.file.FileAlreadyExistsException:
/home/murphy/elastic/elasticsearch/plugins/discovery-gce/build/testrun/test/J0/temp/org.elasticsearch.discovery.gce.GceDiscoverTests_3D8ACCBEE4BD7E73-001/tempDir-002/0
   >     at
org.elasticsearch.discovery.gce.GceDiscoverTests.nodeSettings(GceDiscoverTests.java:90)
   >     at
org.elasticsearch.test.ESIntegTestCase$2.nodeSettings(ESIntegTestCase.java:1839)
   >     at
org.elasticsearch.test.InternalTestCluster.getSettings(InternalTestCluster.java:365)
   >     at
org.elasticsearch.test.InternalTestCluster.buildNode(InternalTestCluster.java:579)
   >     at
org.elasticsearch.test.InternalTestCluster.buildNode(InternalTestCluster.java:562)
   >     at
org.elasticsearch.test.InternalTestCluster.getOrBuildRandomNode(InternalTestCluster.java:468)
   >     at
org.elasticsearch.test.InternalTestCluster.client(InternalTestCluster.java:648)
   >     at
org.elasticsearch.test.ESIntegTestCase.client(ESIntegTestCase.java:641)
   >     at
org.elasticsearch.test.ESIntegTestCase.client(ESIntegTestCase.java:634)
   >     at
org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:553)
   >     at
org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2058)
   >     at java.lang.Thread.run(Thread.java:745)
   > Caused by: java.nio.file.FileAlreadyExistsException:
/home/murphy/elastic/elasticsearch/plugins/discovery-gce/build/testrun/test/J0/temp/org.elasticsearch.discovery.gce.GceDiscoverTests_3D8ACCBEE4BD7E73-001/tempDir-002/0
   >     at
sun.nio.fs.UnixException.translateToIOException(UnixException.java:88)
   >     at
sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
   >     at
sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
   >     at
sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
   >     at
org.apache.lucene.mockfile.FilterFileSystemProvider.createDirectory(FilterFileSystemProvider.java:132)
   >     at
org.apache.lucene.mockfile.FilterFileSystemProvider.createDirectory(FilterFileSystemProvider.java:132)
   >     at
org.apache.lucene.mockfile.FilterFileSystemProvider.createDirectory(FilterFileSystemProvider.java:132)
   >     at
org.apache.lucene.mockfile.FilterFileSystemProvider.createDirectory(FilterFileSystemProvider.java:132)
   >     at java.nio.file.Files.createDirectory(Files.java:674)
   >     at
org.elasticsearch.discovery.gce.GceDiscoverTests.nodeSettings(GceDiscoverTests.java:88)
   >     ... 45 more
  2> NOTE: leaving temporary files on disk at:
/home/murphy/elastic/elasticsearch/plugins/discovery-gce/build/testrun/test/J0/temp/org.elasticsearch.discovery.gce.GceDiscoverTests_3D8ACCBEE4BD7E73-001
  2> NOTE: test params are: codec=Asserting(Lucene70): {}, docValues:{},
maxPointsInLeafNode=1020, maxMBSortInHeap=6.475715560444328,
sim=RandomSimilarity(queryNorm=false): {}, locale=ar-MA,
timezone=Africa/Nouakchott
  2> NOTE: Linux 4.4.0-78-generic amd64/Oracle Corporation 1.8.0_111
(64-bit)/cpus=4,threads=1,free=473409288,total=514850816
  2> NOTE: All tests run in this JVM: [GceDiscoverTests]
Completed [3/4] on J0 in 2.98s, 1 test, 1 error <<< FAILURES!
Tests with failures:
  - org.elasticsearch.discovery.gce.GceDiscoverTests.testJoin
JUnit4 test failed, ant output was:
   [junit4] <JUnit4> says kaixo! Master seed: 3D8ACCBEE4BD7E73
   [junit4] JVM J0:     0.67 ..     4.72 =     4.05s
   [junit4] JVM J1:     0.37 ..     3.42 =     3.05s
   [junit4] JVM J2:     0.37 ..     2.95 =     2.59s
   [junit4] Execution time total: 4.73 sec.
   [junit4] Tests summary: 4 suites, 22 tests, 1 error
:plugins:discovery-gce:test FAILED
:plugins:discovery-gce:integTestCluster#stop
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':plugins:discovery-gce:test'.
> There were test failures: 4 suites, 22 tests, 1 error [seed:
3D8ACCBEE4BD7E73]
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or
--debug option to get more log output.
BUILD FAILED
Total time: 18.286 secs
There were test failures: 4 suites, 22 tests, 1 error [seed:
3D8ACCBEE4BD7E73]

**Comment by rjernst** [comment]
The output explains what is happening:

```
ERROR   1.08s J0 | GceDiscoverTests.testJoin <<< FAILURES!
 Throwable #1: java.lang.IllegalArgumentException: Unknown discovery type [gce]
```

GceDiscoveryTests is an ESIntegTestCase. You need to update the nodeSettings() method to not set the `discovery.type` (which is the old way that your change is removing support for), and instead set `discovery.zen.hosts_provider`.

**Comment by jmorph99** [comment]
Thanks, that did it.

The check against the code base went fine.  
I committed the changes, rebased, pushed, and signed the agreement and then made a pull request on Github site.  If anyone could check to see if I did it correctly that would be good, this is my first time.

John

**Comment by bleskes** [comment]
@jmorph99 I think something went wrong and you made your pull request against your own fork. Can you try again and open it against this one?

**Comment by jmorph99** [comment]
Ok, try it tomorrow.

**Comment by jmorph99** [comment]
Ok, I'm a little confused with your push procedure.
What is the purpose of forking the project?
Should I just start over from the root project?

**Comment by bleskes** [comment]
@jmorph99 maybe this can help: https://help.github.com/articles/creating-a-pull-request/

You basically make the changes on a branch in your fork. Then go there in the github ui and hit the new pull request button.

**Comment by jmorph99** [comment]
I thought that is what I did!

**Comment by jmorph99** [comment]
Ok - try it again and keep your fingers crossed.
